### ISSUE TITLE
FNX-453 ⁃ [Meta] Basic Login Management

### ISSUE THREAD

**Issue by vesta0** [mozilla-mobile/fenix#1284]
### As a user, I want to be able to 
* choose whether I want my browser to remember my logins
* my browser to auto-fill the login fields when I revisit a site
* be able to view my saved logins in my browser settings
* be able to copy my login to clipboard
* be able to search my logins
* be able to delete my logins
* choose whether I want my browser to access my existing synced logins and sync new Fenix logins with my Firefox account 

### Acceptance Criteria
GeckoView Meta Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1575401
Project board for Logins: https://github.com/mozilla-mobile/fenix/projects/22
Advanced Login Management: #5774 

Existing GV Autofill API
[Autofill.Delegate docs](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/Autofill.Delegate.html#onAutofill-org.mozilla.geckoview.GeckoSession-int-org.mozilla.geckoview.Autofill.Node-)
[GeckoSession.autofill docs](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/GeckoSession.html#autofill-android.util.SparseArray-)

Waiting Work:
Login Storage Gecko Toolkit Implementation 
[Bug 1596660 - Create a GeckoView nsILoginManagerStorage backend](https://bugzilla.mozilla.org/show_bug.cgi?id=1596660)

Login Storage GV API Implementation
[Bug 1593069 - Implement the Login Storage API](https://bugzilla.mozilla.org/show_bug.cgi?id=1593069)

* [x] #5074 Sync logins from FxA 
* [x] #5544 View saved logins
* [x] #7087 Search logins
* [x] #5891 Copy a saved login 
* [x] #5545 In settings, I can choose whether I want my browser to remember my logins
* [x] #5542 Saving logins on website
* [x] #5938 Use Android Secure storage for API 23+ for sync keys
* [x] #5550 Choose secure storage library for 21-23
* [x] #5543 AC API for integrating upcoming GV Login Autofill Delegate with Logins Storage
* [x] #6696 Integrate AC API for Autofill Logins Delegate (Land Autofill in Fenix)
* [x] #5546 Setting in login management that links to Sync settings (to sync logins)
* [x] #5547 Add top-level authentication when accessing any passwords
* [x] #6055 Spike for deciding user experience for protection of logins
* [x] #5586 Telemetry for Logins
* [x] #7796 Turn on Logins for release

### UX Spec
[Full mocks](https://share.goabstract.com/d3008a75-3706-4f40-ae81-e4c443062d2a?mode=design&sha=35d2d93e5707e7c1c79d0aba3613d5cf1340e6f4)

**Comment by cpeterso** [comment]
@vesta0 - Is password management blocked on GV? Doesn't password autofill already work (excluding some bugs) with third-party password manager apps like Lockwise (Fenix issue #87)?

**Comment by NatoBoram** [comment]
Third-party password managers or autofills don't work in Android pre-Oreo.

**Comment by liuche** [comment]
We should carve out some time in an upcoming sprint to figure out the dependencies, as well as Form autofill. The goal of that investigation would be to list (GeckoView) dependencies, come up with a few scopes of issues.

This scoping and investigation should be done in Q2.
@vesta0 will file an investigation issue.

**Comment by liuche** [comment]
Will have a kick off meeting next week, with A-S people included. Vesta is working on the feature brief.

**Comment by AmyYLee** [comment]
Sizing this as M

**Comment by mnoorenberghe** [comment]
> Will have a kick off meeting next week, with A-S people included.

Can I be included in that meeting as a maintainer of Lockwise/Password Manager and Form Autofill on desktop Fx.?

**Comment by vesta0** [comment]
Hey @mnoorenberghe I will be sure to include you in our next sync up!

**Comment by cpeterso** [comment]
This issue has the `needs:gv` label. What GV work is still needed for login management? Form autofill bugs?

**Comment by mnoorenberghe** [comment]
> This issue has the `needs:gv` label. What GV work is still needed for login management? 

That probably depends on how this will be implemented… one approach would be to expose the existing Fennec about:logins page which would be a GV change.

> Form autofill bugs?

Those wouldn't fall under management IMO. Though I guess "Have my browser auto-fill the login fields when I revisit a site" is part of this user story… Btw. "Form autofill" usually refers to addresses and credit cards, not logins.

**Comment by ekager** [comment]
I opened https://bugzilla.mozilla.org/show_bug.cgi?id=1575401 for discussion about any GV work and maybe alternative options we have here.

**Comment by cpeterso** [comment]
The GV team closed GV bug 1575401 as WONTFIX because snorp says GV already has the necessary API: GV's SessionTextInput API. For Android O+ you would still probably want to use a system autofill provider. Perhaps A-C can create an autofill abstraction for Fenix that encapsulates both GV's SessionTextInput API and Oreo's autofill API, depending on the OS version at run time?

**Comment by mnoorenberghe** [comment]
Part of the reason I wanted to meet is because there are various pieces to a password manager that get forgotten or conflated. Some of the key pieces are:

* Login field heuristics
  * GV has some very basic support for this using the Toolkit password manager code
* Autofill (without interaction) and autocomplete (with a list of relevant logins)
  * This includes support for per-site overrides recipes
* Capture/saving of submitted logins (including heuristics and UI)
  * This should also handle not capturing non-login data in password fields e.g. credit card numbers.
  * This includes support for per-site overrides recipes.
* Management UI
  * Ideally shared with Lockwise or using the Fennec HTML about:logins page in the short term
* Storage/Sync
  * Either using the existing toolkit storage mozStorage/SQLite backend that already works with autofill or using the Rust login a-c which is used on FxiOS and in Lockwise apps)
* Protection of the saved data
  * A master password and/or biometric authentication before revealing/using a saved login.

These are just some of the pieces off the top of my head. All of the above already existed in Fennec (mostly via the Toolkit Password Manager code) btw.

Password generation and breached login warnings were also added to Fx70 on the desktop and would be good to ship on Fenix eventually. Building on top of Toolkit means not having to re-implement every improvement desktop gets.

I think the first step should be choosing a storage approach that will work with both management UI and existing capture/fill logic then figuring out the milestones and scoping.

**Comment by lime124** [comment]
right now on track for this to wrap up next week.

**Comment by bifleming** [comment]
There is no additional A-S work to be done, per Janet.

**Comment by liuche** [comment]
We want to have this broken down into individual issues, and have them sized (or ready to be sized), by Wed 9/25. @lime124 @ekager @liuche

**Comment by liuche** [comment]
When Amy gets back, she'll finish off the Phase 1 work (finishing this mock).
@vesta0 will file a new issue for Phase 2 (integrating lockwise branding, #5551 multiple logins, whatever is missing from the current Acceptance Criteria list in comment 0)

**Comment by AmyYLee** [comment]
@ekager @vesta0 The mocks have been updated and ready for engineering. Follow-up: Review with Legal what are our options for fallback if a user doesn't have a PIN set and biometrics has failed. Do we still show full login/password credentials?

**Comment by liuche** [comment]
@AmyYLee here's the spike where the outcomes for these fallback options should be documented #6055

**Comment by liuche** [comment]
@cpeterso it looks like the only remaining GV bug against the meta is https://bugzilla.mozilla.org/show_bug.cgi?id=1577003

Is that correct, or are there other bugs that should be blocking the Logins Meta? https://bugzilla.mozilla.org/show_bug.cgi?id=1575401

If it is the only outstanding GV item, once it lands, does it allow Fenix to begin implementing Saving Logins #5542 and Autofilling Logins #5543?

**Comment by cpeterso** [comment]
> @cpeterso it looks like the only remaining GV bug against the meta is https://bugzilla.mozilla.org/show_bug.cgi?id=1577003

Do you mean bug https://bugzilla.mozilla.org/show_bug.cgi?id=1577005? Bug 1577003 was fixed two weeks ago.

Bug 1577005 is the last known blocking bug and it has a patch up for review. It's needed to create or update saved logins in Gecko's autofill system (Fenix #5542).

You can start implementing Autofilling Logins #5543 now, though Gecko will have no saved logins to autofill until bug 1577005 and #5542 are fixed.

**Comment by ekager** [comment]
Realized this bug hasn't been udpated with new bugs coming from GV/Gecko toolkit/storage discussions: 

Existing GV Autofill API
[Autofill.Delegate docs](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/Autofill.Delegate.html#onAutofill-org.mozilla.geckoview.GeckoSession-int-org.mozilla.geckoview.Autofill.Node-)
[GeckoSession.autofill docs](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/GeckoSession.html#autofill-android.util.SparseArray-)

Future Work:
Login Storage Gecko Toolkit Implementation 
[Bug 1596660 - Create a GeckoView nsILoginManagerStorage backend](https://bugzilla.mozilla.org/show_bug.cgi?id=1596660)

Login Storage GV API Implementation
[Bug 1593069 - Implement the Login Storage API](https://bugzilla.mozilla.org/show_bug.cgi?id=1593069)

We have an idea of what the API will look like so basic AC work can be started, but these bugs will need to land in the next few weeks before autofill/capture can land in Fenix!

**Comment by ekager** [comment]
Closing this because everything except search is done and we have https://github.com/mozilla-mobile/fenix/issues/7087 open for search
